### PR TITLE
DEPR: deprecate string H, BH, CBH in offsets frequencies, resolution abbreviations, _attrname_to_abbrevs

### PR DESCRIPTION
**Pull request by natmokval**
xref #54061

- deprecated for Timedelta units 'H' in favour of 'h'
- deprecated for time series frequencies aliases 'H', 'BH', 'CBH' in favour of 'h', 'bh', 'cbh'
- deprecated resolution 'H' for Timedelta.resolution_string in favour of 'h'

### CODE REVIEW

File: pandas/_libs/tslibs/offsets.pyx [hunk 1]
```diff
@@ -2014,7 +2014,7 @@ cdef class BusinessHour(BusinessMixin):
             nb_offset = 1
         else:
             nb_offset = -1
-        if self._prefix.startswith("C"):
+        if self._prefix.startswith(("C", "c")):
             # CustomBusinessHour
             return CustomBusinessDay(
```
Comment (MarcoGorelli) [inline pandas/_libs/tslibs/offsets.pyx:2019]: should this one check for both? if it's business hour, don't we know it's going to start with `'c'`?

File: pandas/_libs/tslibs/offsets.pyx [hunk 2]
```diff
@@ -2174,7 +2174,7 @@ cdef class BusinessHour(BusinessMixin):
 
         # adjust by business days first
         if bd != 0:
-            if self._prefix.startswith("C"):
+            if self._prefix.startswith("c"):
                 # GH#30593 this is a Custom offset
                 skip_bd = CustomBusinessDay(
```
Comment (MarcoGorelli) [inline pandas/_libs/tslibs/offsets.pyx:2179]: as above, if we know if it's custom business day, won't it be `'C'`?

File: pandas/tests/indexes/timedeltas/test_constructors.py [hunk 3]
```diff
@@ -175,10 +175,11 @@ def test_constructor(self):
...
         # GH #21877
-        expected = timedelta_range("1s", periods=9, freq="s")
-        durations = [f"P0DT0H0M{i}S" for i in range(1, 10)]
-        result = to_timedelta(durations)
-        tm.assert_index_equal(result, expected)
+        timedelta_range("1s", periods=9, freq="s")
+        [f"P0DT0H0M{i}S" for i in range(1, 10)]
+        to_timedelta("P0DT0H0M1S")
+        # result = to_timedelta(durations)
+        # tm.assert_index_equal(result, expected)
```
Comment (MarcoGorelli) [inline pandas/tests/indexes/timedeltas/test_constructors.py:182]: why this change?

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]
thanks for working on this - got a couple of minor comments, but this looks like it's pretty close

File: pandas/tests/indexes/timedeltas/test_constructors.py (hunk 3 above)
Comment (natmokval) [reply pandas/tests/indexes/timedeltas/test_constructors.py]: oops, thanks for noticing this. I changed it while I was working on fixing tests. I rolled my changes back.

File: pandas/_libs/tslibs/offsets.pyx (hunk 1 above)
Comment (natmokval) [reply pandas/_libs/tslibs/offsets.pyx:2019]: Thank you for reviewing this PR.

I think we need to check both: upper and lowercase 'c', because we added `"cbh"` to the dictionary _`dont_uppercase = {"h", "bh", "cbh", "MS", "ms", "s", "me"}.` 
On the other hand, if freq is not in  `_dont_uppercase`  we do uppercase in `_get_offset()`.

When I remove the check for "c", tests for CustomBusinessHour fail.

File: pandas/_libs/tslibs/offsets.pyx (hunk 2 above)
Comment (natmokval) [reply pandas/_libs/tslibs/offsets.pyx:2179]: I'm not sure, is it always custom business day? Some tests for `CustomBusinessHour` , e.g.
`test_custom_businesshour_weekmask_and_holidays`
`test_us_federal_holiday_with_datetime`
go to line 2179, , which is why I changed capitalisation.

**Comment by natmokval** [comment]
I added tests for Hour and CustomBusinessHour and corrected the definition of `to_offset`. @MarcoGorelli, could you please take a look at the update?

File: pandas/_libs/tslibs/offsets.pyx (hunk 1 above)
Comment (MarcoGorelli) [reply pandas/_libs/tslibs/offsets.pyx:2019]: thanks for explaining - the comment 
```python
            # CustomBusinessHour
            return CustomBusinessDay
```
might be a bit misleading, but I'll pull your code tomorrow, step through it, and check

the rest looks good though, so we should be very close

File: pandas/_libs/tslibs/offsets.pyx (hunk 2 above)
Comment (MarcoGorelli) [reply pandas/_libs/tslibs/offsets.pyx:2179]: right sorry, this is `CustomBusinessHour`, so `'c'` should always be sufficient

all good then, resolving this comment

File: pandas/_libs/tslibs/offsets.pyx (hunk 1 above)
Comment (natmokval) [reply pandas/_libs/tslibs/offsets.pyx]: Thanks, I agree, checking `if self._prefix.startswith("c"):` should be enough. I changed the condition, locally all tests passed.

**Review by MarcoGorelli — COMMENTED** [review]
looks good, just a couple of things I don't quite understand

File: pandas/_libs/tslibs/offsets.pyx [hunk 4]
```diff
@@ -4660,8 +4651,19 @@ cpdef to_offset(freq, bint is_period=False):
...
+                    if prefix in c_DEPR_ABBREVS:
+                        warnings.warn(
+                            f"\'{prefix}\' is deprecated and will be removed "
+                            f"in a future version. Please use "
+                            f"\'{c_DEPR_ABBREVS.get(prefix)}\' "
+                            f"instead of \'{prefix}\'.",
+                            FutureWarning,
+                            stacklevel=find_stack_level(),
+                        )
+                        prefix = c_DEPR_ABBREVS[prefix]
```
[the remark this answers is not in the document]
Comment (natmokval) [reply pandas/_libs/tslibs/offsets.pyx:4663]: I moved the check and transformation `prefix = c_DEPR_ABBREVS[prefix]` to keep it together with the call  `_get_offset` in line 4666, where we use the new prefix.
I think it's not necessary, I can move it back.

File: pandas/_libs/tslibs/offsets.pyx [hunk 5]
```diff
@@ -4660,8 +4651,19 @@ cpdef to_offset(freq, bint is_period=False):
...
+
                     stride = int(stride)
-                    offset = _get_offset(name)
+                    offset = _get_offset(prefix)
```
[the remark this answers is not in the document]
Comment (natmokval) [reply pandas/_libs/tslibs/offsets.pyx:4666]: we had tests that didn't work for BH / CBH because we changed prefix to bh / cbh ` in line 4663 prefix = c_DEPR_ABBREVS[prefix]`, but then passed to `_get_offset` `name,` = `BH / CBH`. That is why I replaced name with prefix.

File: pandas/_libs/tslibs/offsets.pyx (hunk 4 above)
[the remark this answers is not in the document]
Comment (natmokval) [reply pandas/_libs/tslibs/offsets.pyx]: It was my mistake, I moved the check `if prefix in c_DEPR_ABBREVS:` back. Seems like we don't have a test to cover this case ,I added a test in PR "DEPR: 'A' in favour of 'Y'"  that fails if the order of checks isn't correct.

**Comment by MarcoGorelli** [comment]
looks like we're close, just got some conflicts from the A -> Y one (but I think you've mastered resolving conflicts now, so should be fine 😄 )

**Comment by natmokval** [comment]
> looks like we're close, just got some conflicts from the A -> Y one (but I think you've mastered resolving conflicts now, so should be fine 😄 )

thanks, I resolved conflicts, ci is green.

**Review by MarcoGorelli — APPROVED** [review]
looks solid, nice one!
